aten/src/ATen/native/TensorIndexing.cpp

```diff
@@ -154,5 +152,189 @@ std::ostream& operator<<(std::ostream& stream, const std::vector<TensorIndex>& t
   return stream;
 }
 
+// This mirrors `count_specified_dimensions` in torch/csrc/autograd/python_variable_indexing.cpp
+int64_t count_specified_dimensions(ArrayRef<TensorIndex> indices) {
+  // Count the number of indexed dimensions (everything but ellipsis and None)
+  int64_t count = 0;
+  size_t size = indices.size();
+  for (size_t i = 0; i < size; i++) {
+    auto& obj = indices[i];
+    if (obj.is_tensor()) {
+      auto& tensor = obj.tensor();
+      if (tensor.scalar_type() == kByte || tensor.scalar_type() == kBool) {
+        count += tensor.dim();
+      } else {
+        count++;
+      }
+    } else if (!obj.is_none() && !obj.is_ellipsis() && !obj.is_boolean()) {
+      count++;
+    }
+  }
+  return count;
+}
+
+// This mirrors `applySlicing` in torch/csrc/autograd/python_variable_indexing.cpp
+Tensor applySlicing(const Tensor& self, ArrayRef<TensorIndex> indices, std::vector<Tensor>& outIndices) {
+  int64_t size = indices.size();
+  int64_t dim = 0;
+  int64_t specified_dims = count_specified_dimensions(indices);
+
+  TORCH_CHECK_INDEX(specified_dims <= self.dim(), "too many indices for tensor of dimension ", (int)self.dim());
+
+  Tensor result = self;
+  for (int64_t i = 0; i < size; i++) {
+    auto& obj = indices[i];
+    if (obj.is_integer()) {
+      result = handleInteger(result, dim, obj.integer(), i);
+    } else if (obj.is_slice()) {
+      result = handleSlice(result, dim, obj.slice().start(), obj.slice().stop(), obj.slice().step());
```

Comment:
More on the mutable reference situation. As a reviewer, I may have very reasonably asked, while comparing the new and the old code: "What happened to the dim increment?!"